FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/pkcs11_operations.c

```diff
@@ -0,0 +1,1038 @@
+        privateKeyTemplate[ 4 ].pValue = &xTrueObject;
+
+        xResult = xFunctionList->C_CreateObject( xSession,
+                                                 ( CK_ATTRIBUTE_PTR ) &privateKeyTemplate,
+                                                 sizeof( privateKeyTemplate ) / sizeof( CK_ATTRIBUTE ),
+                                                 &xObjectHandle );
+    }
+
+    if( NULL != pxRsaParams )
+    {
+        free( pxRsaParams );
+    }
+
+    return xResult;
+}
+
+/*-----------------------------------------------------------*/
+
+static CK_RV prvProvisionPrivateKey( CK_SESSION_HANDLE xSession,
+                                     const char * pcPrivateKey,
+                                     size_t xPrivateKeyLength,
+                                     const char * pcLabel )
+{
+    CK_RV xResult = CKR_OK;
+    mbedtls_pk_type_t xMbedKeyType = MBEDTLS_PK_NONE;
+    int xMbedResult = 0;
+    mbedtls_pk_context xMbedPkContext = { 0 };
+
+    mbedtls_pk_init( &xMbedPkContext );
+    xMbedResult = mbedtls_pk_parse_key( &xMbedPkContext, ( const uint8_t * ) pcPrivateKey,
+                                        xPrivateKeyLength, NULL, 0 );
+
+    if( xMbedResult != 0 )
+    {
+        LogError( ( "Unable to parse private key." ) );
+        xResult = CKR_ARGUMENTS_BAD;
+    }
+
+    /* Determine whether the key to be imported is RSA or EC. */
+    if( xResult == CKR_OK )
+    {
+        xMbedKeyType = mbedtls_pk_get_type( &xMbedPkContext );
+
+        if( xMbedKeyType == MBEDTLS_PK_RSA )
+        {
+            xResult = prvProvisionPrivateRSAKey( xSession, pcLabel, &xMbedPkContext );
+        }
+        else if( ( xMbedKeyType == MBEDTLS_PK_ECDSA ) ||
+                 ( xMbedKeyType == MBEDTLS_PK_ECKEY ) ||
+                 ( xMbedKeyType == MBEDTLS_PK_ECKEY_DH ) )
+        {
+            xResult = prvProvisionPrivateECKey( xSession, pcLabel, &xMbedPkContext );
+        }
+        else
+        {
+            LogError( ( "Invalid private key type provided. Only RSA-2048 and "
+                        "EC P-256 keys are supported." ) );
+            xResult = CKR_ARGUMENTS_BAD;
+        }
+    }
+
+    mbedtls_pk_free( &xMbedPkContext );
+
+    return xResult;
+}
+
+/*-----------------------------------------------------------*/
+
+static CK_RV prvProvisionCertificate( CK_SESSION_HANDLE xSession,
+                                      const char * pcCertificate,
+                                      size_t xCertificateLength,
+                                      const char * pcLabel )
+{
+    PKCS11_CertificateTemplate_t xCertificateTemplate;
+    CK_OBJECT_CLASS xCertificateClass = CKO_CERTIFICATE;
+    CK_CERTIFICATE_TYPE xCertificateType = CKC_X_509;
+    CK_FUNCTION_LIST_PTR xFunctionList = NULL;
+    CK_RV xResult = CKR_OK;
+    uint8_t * pucDerObject = NULL;
+    int32_t ulConversion = 0;
+    size_t xDerLen = 0;
+    CK_BBOOL xTokenStorage = CK_TRUE;
+    CK_BYTE pxSubject[] = "TestSubject";
```

Comment:
This should be the DER encoding of the certificate subject, not an arbitrary string.

Reply:
See http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/os/pkcs11-base-v2.40-os.html section 4.6.3

Reply:
Resolved in chat - requires loading the cert into mbedtls to extract the subject